### PR TITLE
Issue #213 - UI - Initial support for ActivityPub comments

### LINKED ISSUE #213: Add support for comments
## Dependencies
- the content of the socialInteract tag must be returned by the API
    - https://github.com/Podcastindex-org/docs-api/issues/91 

## Remarks
- Might be better if we implement a view for individual episodes

## References
- https://www.npmjs.com/package/threadcap
- https://github.com/podStation/podStation/issues/304

### PR DESCRIPTION
**Pull request by dellagustin**
This commit introduces initial support for showing comments. It relies on the API also including socialInteracts and on a new API on the server being implemented to fetch the comments.

~This is not yet ready, I have implemented some mock code on the server side that is not included in this commit.~

The front-end code expects the comments API to respond with the same format as podverse's `/api/v1/episode/{episodeid}/proxy/activity-pub`, which is exactly what is returned by [threadcap](https://www.npmjs.com/package/threadcap).

## To Do
- [x] Sanitize content html - done with https://github.com/Podcastindex-org/web-ui/pull/218/commits/07207d294015adee34f7f7df34c13a7e41a280b5
- [x] Fetch comments on the server using threadcap - done with bdf237e3a60197a6ea09a29873380379ba14793f
- [x] Add state for "loading comments" - currently is taking circa 7s, so there is a noticeable wait - done with https://github.com/Podcastindex-org/web-ui/pull/218/commits/5282362872aa9c6c1c41a2c8153566e5ac92614e
- [x] Solve the react warning _Warning: Each child in a list should have a unique "key" prop._ - done with https://github.com/Podcastindex-org/web-ui/pull/218/commits/22f868b02b5a4944bdf5cf4b36a0da996658fba3
- [x] Define global constant for user agent - done with https://github.com/Podcastindex-org/web-ui/pull/218/commits/2f826a759137971187fff2039be6780efde49168
- [x] Use socialInteract from the API, when it becomes available - done with https://github.com/Podcastindex-org/web-ui/pull/218/commits/f2037c23caadf7b6488eaf0ba26049ef75cf19a9
- [x] Show comment publication date and time - done with https://github.com/Podcastindex-org/web-ui/commit/b48dc1b5c6114406543a0428cc45119801e419cc
- [x] Support language-tagged values - see https://github.com/Podcastindex-org/web-ui/pull/218#issuecomment-1409452414 - done with https://github.com/Podcastindex-org/web-ui/pull/218/commits/608dcfba6efea200ed534a8c9ec7e975d10fe47a
- [x] Implement @nathangathright's  design - https://github.com/Podcastindex-org/web-ui/pull/218#issuecomment-1409912989 - done with - https://github.com/Podcastindex-org/web-ui/pull/218/commits/b48dc1b5c6114406543a0428cc45119801e419cc
- [x] Support `summary` (stretch goal)- see https://github.com/Podcastindex-org/web-ui/pull/218#issuecomment-1411052250 - done with https://github.com/Podcastindex-org/web-ui/pull/218/commits/094393d3cf597dc6515784729b5342e736776a4a
- [x] Fix content overflow issue (stretch goal) - see https://github.com/Podcastindex-org/web-ui/pull/218#issuecomment-1415397578 - done with https://github.com/Podcastindex-org/web-ui/pull/218/commits/d33af68580a1aeb714ca3c5cac309938d21de371 

### Performance
When this PR is ready for merge, it will still be limited to fetching all comments in a single shot, and the UI will wait until they are all fetched, which can take some time.

A potential improvement would be some kind of "long pooling", where the server sends partial updates back each time a new comment is fetched. This way the UI could be progressively updated. Maybe this can be achieved with https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Using_readable_streams.

I have created an issue to track this future improvement: https://github.com/Podcastindex-org/web-ui/issues/228

### Summary / Content warning 
The property `summary` is not taken into account, which is returned by Pleroma and Mastodon posts as a _Content warning_ (e.g. to avoid spoilers). ~This is, to my best knowledge, also a limitation of threadcap (see https://github.com/skymethod/minipub/issues/6)~, not anymore, see https://github.com/Podcastindex-org/web-ui/pull/218#issuecomment-1411052250. I set this now as a _stretch goal_, could be easily added later in a compatible way with a separate PR, if this one is dragging too long.

### Attachments
Attachments are currently not shown (e.g. attached images).

I have created an issue to track this future improvement: https://github.com/Podcastindex-org/web-ui/issues/229.

### Scaling
As all the API calls are going out from the same IP, if many users are checking comments in parallel, they could reach the instances' rate limit. For instance, the default for Mastodon instances is 300 calls within 5 mins ([mastodon rate limits](https://docs.joinmastodon.org/api/rate-limits/#per-account)). 

## Related issues
- resolves https://github.com/Podcastindex-org/web-ui/issues/213

### Dependencies
- the content of the socialInteract tag must be returned by the API
    - https://github.com/Podcastindex-org/docs-api/issues/91 
- (weak dependency) created https://github.com/skymethod/minipub/issues/6
- (weak dependency) created https://github.com/skymethod/minipub/issues/7

## Links and References
- [socialInteract](https://github.com/Podcastindex-org/podcast-namespace/blob/main/docs/1.0.md#social-interact) at the 
- Fediverse threads
    - [Main thread discussing this issue](https://fosstodon.org/@dellagustin/109768865066074877)
    - [Promotion for this issue](https://podcastindex.social/@dave/109775921732686645) by @daveajones   
- https://github.com/podStation/podStation/issues/304
- Activity Streams 2.0 / ActivityPub
    - https://www.w3.org/TR/activitystreams-core/ 
    - https://www.w3.org/TR/activitystreams-vocabulary/
    - https://www.w3.org/TR/activitypub/

### CODE REVIEW

**Comment by dellagustin** [comment]
Additional server side code I'm using at `server/index.js`, to mock the API data:

```js
const fakeCommentsResponse = require('./fake-comments-response.json')
```

```js
app.use('/api/episodes/byfeedid', async (req, res) => {
    let feedId = req.query.id
    let max = req.query.max
    const response = await api.episodesByFeedId(feedId, null, max)

    // Injects fake socialInteract content here for testing
    response.items.forEach((item) => {
        item.socialInteracts = [{
            uri: 'https://podcastindex.social/users/dave/statuses/109683341113064081',
            protocol: 'activitypub',
        }]
    })

    res.send(response)
})
```

```js
// ------------------------------------------------
// ------------ API to get comments for episode ---
// ------------------------------------------------
app.use('/api/comments/byepisodeid', async (req, res) => {
    let feedUrl = req.query.url
    // fakeCommentsResponse was taken directly from the API used by podverse
    res.send(fakeCommentsResponse)
})
```

For `server\fake-comments-response.json` I simply inspected the return of the API in podverse and copied the response body.

Current result:
image

**Comment by dellagustin** [comment]
@daveajones, we'll need to sanitize the content of the comments, do you have any library recommendations?

**Comment by daveajones** [comment]
Threadcap doesn't sanitize?

**Comment by dellagustin** [comment]
> Threadcap doesn't sanitize?
@daveajones I'm not sure, but I am assuming it doesn't. I could not find information indicating whether it does or does not on the documentation. I asked @johnspurlock-skymethod, see https://github.com/skymethod/minipub/issues/5 and https://fosstodon.org/@dellagustin/109771551290816322.

Now that's the server is doing requests to 3rd party servers, we should define a user agent. Do you have any preferences on this one?

I'll move forward with missing pieces, but soon it will be blocked only by https://github.com/Podcastindex-org/docs-api/issues/91.

Could you do a "call to action" for the community to contribute on the "beautification" part? Maybe on the next Podcasting 2.0 episode? This is a part were I'm very weak.

**Comment by dellagustin** [comment]
Current error handling:

image

image

**Comment by daveajones** [comment]
I will 100% do a call out on the next board meeting. For user agent I typically use Podcastindex.org/v1.x so maybe something like:

Podcastindex.org-web/v1.0 ?

**Comment by Clay-Ferguson** [comment]
> @daveajones, we'll need to sanitize the content of the comments, do you have any library recommendations?

I recommend DOMPurify NPM module for sanitizing. I have been using it in my project "Quanta" for a while.

**Comment by dellagustin** [comment]
One current limitation, I'm only taking into account English language content.

threadcap returns the content of the comments as language-tagged strings:
```json
"content": {
          "en": "content in English"
        }
```
I'm currently hardcoding to English.

Note that the content on the Activity is not always language-tagged, and when it is, this is under the property `contentMap` instead of `content`.

e.g.:
```json
"content": "content in English",
    "contentMap": {
        "en": "content in English"
    },
```

I was remembered about this when testing with a reply from a different account where it is not set to English (simplified example below):

```json
{
    "@context": [
        "https://www.w3.org/ns/activitystreams",
        "https://blob.cat/schemas/litepub-0.1.jsonld",
        {
            "@language": "und"
        }
    ],
    "content": "Content in und",
}
```

Note that in the absence of a `contentMap` threadcap seems to be building one from `content` and `@language`.

**Comment by dellagustin** [comment]
@daveajones 
> Podcastindex.org-web/v1.0 ?

Sounds good to me. Is it ok if I read the version from `package.json` to build the final user agent?

**Comment by dellagustin** [comment]
Design by @nathangathright: https://codepen.io/nathangath/pen/jOpvKJo?editors=1000

image

**Comment by nathangathright** [comment]
CleanShot 2023-01-31 at 09 07 54

I’ve just updated the CodePen with a custom disclosure chevron, better support for Safari, and some space-saving mobile enhancements.

**Comment by dellagustin** [comment]
@nathangathright my mouse pointer is going a bit haywire on the clickable elements of the UI at the current version, but it may be a codepen thing, let's see if it happens when I implement it in PI.
It shows a hand cursor as if it was clickable, but if I let it there it goest back to being an arrow.

**Comment by dellagustin** [comment]
So an expectation management message: as you can see, the To Do list is packed. What I did so far was during a "time multiplexing window" that I dedicate to this project, now I have to do the same to other projects 😄 . Also, I have the feeling we are at a "80% of the functionality with 20% of the effor" point. In other words, the what is left may take more time then you would expect to be ready. I'm guessing something like 4 weeks we me "multiplexing" between projects and life. Let's see.

**Comment by johnspurlock-skymethod** [comment]
> The property summary is not taken into account, which is returned by Pleroma and Mastodon posts as a Content warning (e.g. to avoid spoilers). This is, to my best knowledge, also a limitation of threadcap (see https://github.com/skymethod/minipub/issues/6).

Just for you Guilherme, this morning I added `summary` to the comment payload in the latest version of [Threadcap](https://www.npmjs.com/package/threadcap) - version 0.1.10

**Comment by nathangathright** [comment]
@dellagustin Yeah, `cursor:pointer` on absolutely-positioned pseudo-elements might not be the best idea. Given that [all interactive elements don't need a hand cursor](https://adamsilver.io/blog/buttons-shouldnt-have-a-hand-cursor/#the-pointer-cursor-is-for-links), I’ll remove that affordance.

I’ve pushed an updated code to a stand-alone repo with GitHub Pages so you can see the behavior without CodePen’s interference. Additionally, I’ve mocked up what Summarized/Content Warnings could look like.

https://nathangathright.github.io/comment-thread-mockup/

CleanShot 2023-01-31 at 16 43 16

**Comment by nathangathright** [comment]
> I'll probably leave the `<relative-time>` usage out of this PR as it adds another dependency, if I understood it correctly, and implement it afterwards with a dedicated PR. Is this this the one you used, https://github.com/github/relative-time-element?

Yeah, that’s fine. Just replace that custom element with the HTML5 standard `[<time>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time)` element. And yes, I just used unpkg to include GitHub’s package as a module at the bottom of the page, and everything else took care of itself.

**Comment by dellagustin** [comment]
@johnspurlock-skymethod , @nathangathright you both have to slow down on your great and timely contributions, it is making me look bad 😛.
Seriously now, the teamwork here is amazing, open source Christmas came early this year for me. I cannot wait to get this finished and integrated, it will be awesome!

**Comment by dellagustin** [comment]
@nathangathright , one note about the _Content Warning_, event though that the semantic for _summary_ used by (at least) _Mastodon_ and _Pleroma_, they show this in a very "neutral" way, meaning: they call it Content Warning when your are editing, but they don't refer to it when they are showing it, they just hide the content under a _Show More_/_Show Content_ user interaction. Screenshots below.
I like this neutral approach in the context of ActivityPub, as the [Activity Vocabulary](https://www.w3.org/TR/activitystreams-vocabulary/) does not define a way to represent content warnings or spoiler alerts (that I could find), so _summary_ is intended to be just a summary, and the neutra approach does not go against that.
In summary (😜), I suggest our implementation support summary in a similar way, just show the summary and have a _Show more_ or _Show content_ button.

image
image

image
image

**Comment by dellagustin** [comment]
Testing with threadcap v0.1.10 showed a regression, reported with https://github.com/skymethod/minipub/issues/7.

**Comment by nathangathright** [comment]
Fair enough, I was copying Ivory, but happy to make it more neutral. 

CleanShot 2023-02-02 at 12 06 05

**Comment by dellagustin** [comment]
> Fair enough, I was copying Ivory, but happy to make it more neutral. 
> 
> CleanShot 2023-02-02 at 12 06 05
> 

@nathangathright perfect!

**Comment by johnspurlock-skymethod** [comment]
> Testing with threadcap v0.1.10 showed a regression

Just published threadcap [0.1.11](https://www.npmjs.com/package/threadcap/v/0.1.11) that should fix it

**Comment by dellagustin** [comment]
The last update on threadcap fixed the issue introduced with v.1.10.
I'm in the works with @nathangathright's design, first version implemented with https://github.com/Podcastindex-org/web-ui/pull/218/commits/b48dc1b5c6114406543a0428cc45119801e419cc
image

**Comment by johnspurlock-skymethod** [comment]
One UI suggestion: remove the leading usernames (aka the twitter canoe) unless they are new to the comment node stack (ie they are added)

I did this in an old site I made called [Convopage](https://convopage.com/c/1310566897541406720) and it makes the comment text _much_ more readable

**Comment by dellagustin** [comment]
> One UI suggestion: remove the leading usernames (aka the twitter canoe) unless they are new to the comment node stack (ie they are added)
> 
> I did this in an old site I made called [Convopage](https://convopage.com/c/1310566897541406720) and it makes the comment text _much_ more readable

I'm not sure how I feel about that, but I'll consider it - not likely in this PR though.

**Comment by dellagustin** [comment]
@nathangathright I can't post a screenshot of the issue right now, but I tested the code you contributed in mobile / responsive mode with the browser and the content is overflowing over the right border.
If you pull the code for this PR for testing, you need to modify the file [`server/index.js`](https://github.com/podStation/podcastindex-web-ui/blob/b48dc1b5c6114406543a0428cc45119801e419cc/server/index.js#L105-L110) as follows (I have not pushed this as it is temp code that will not be needed once https://github.com/Podcastindex-org/docs-api/issues/91 is done).
```js
app.use('/api/episodes/byfeedid', async (req, res) => {
    let feedId = req.query.id
    let max = req.query.max
    const response = await api.episodesByFeedId(feedId, null, max)

    // Injects fake socialInteract content here for testing
    response.items.forEach((item) => {
        item.socialInteracts = [{
            uri: 'https://podcastindex.social/users/dave/statuses/109683341113064081',
            protocol: 'activitypub',
        }]
    })

    res.send(response)
})
```

image

**Comment by dellagustin** [comment]
@nathangathright , I just implemented the support for `summary`. When I saw that your css trickery saved me from managing the state of Show More / Less, I was relieved!

**Comment by dellagustin** [comment]
I have been thinking about how stream back the comments processed by threadcap to the front-end, so that the user does not need to wait for the threadcap to finish its work to start seeing comments.

The idea is to use the threadcap callback and send back a JSON with **new** nodes and and commenters each time there is a `node-processed` event for a comment, then rebuilt the response on the frontend based on these chunked updates and update the state each time a chunk arrives.

**Comment by dellagustin** [comment]
Indication that comments are loading is there: 
image
Ugly, but it works 😄 
@nathangathright , if you feel like beautifying that, come along.
Also, did you see my overflow message from earlier - https://github.com/Podcastindex-org/web-ui/pull/218#issuecomment-1415397578?
One more TODO to go and the blockers from my side would be gone - I'm racing @daveajones here, hoping to finish this before he finishes introducing socialInteract to the API 😜.

**Comment by nathangathright** [comment]
@dellagustin The CSS for the summary element didn't account for display names + handles + long timestamps. Adding `width:100%; overflow:hidden;` on that element would patch the problem. 

But I’ve gone a bit further and [pushed an update](https://github.com/nathangathright/comment-thread-mockup/commit/e02cbe785b01ea033fc50b36a31f9d2952780a9f) to my little repo with stacked name and handles on desktop and hiding the handles on mobile.

(I still think handles are unnecessary given that you can click through to the original profile to verify anyone’s identity.)

**Comment by dellagustin** [comment]
Hello @nathangathright , thank you for the changes. I can see they work well in your standalone mock up, but when I implemented them inside the podcast index, it still overflows, here is a screenshot:
image

**Comment by dellagustin** [comment]
Fixed it by adding width: 100% to an outer div I added with the comments section.
I did remove the body CSS that came with @nathangathright 's mock up, as I feared it would affect the rest of the site somehow.

**Comment by dellagustin** [comment]
Content sanitization added with https://github.com/Podcastindex-org/web-ui/pull/218/commits/07207d294015adee34f7f7df34c13a7e41a280b5 using DOMPurify.

To test it, I faked the response of `/api/comments/byepisodeid`
with a hard coded json file containing in the content of comments
some examples from the section "Some purification samples please?".

References:
- https://www.npmjs.com/package/dompurify

**Comment by dellagustin** [comment]
@daveajones , ready for review 🎉 .

File: server/index.js [hunk 1]
```diff
@@ -112,6 +118,32 @@ app.use('/api/add/byfeedurl', async (req, res) => {
...
+// ------------------------------------------------
+// ------------ API to get comments for episode ---
+// ------------------------------------------------
+app.use('/api/comments/byepisodeid', async (req, res) => {
+    let episodeId = req.query.id;
+    const response = await api.episodesById(episodeId, false);
+
+    const socialInteract = response.episode.socialInteract && response.episode.socialInteract.filter((si) => si.protocol === 'activitypub');
+
+    if(!socialInteract && socialInteract.lenght >= 0) {
```
Comment (ericpp) [inline server/index.js:130]: @dellagustin FYI. It looks like the length property got typoed here ( `.lenght` )

Comment (dellagustin) [reply server/index.js:130]: @ericpp good catch! I thought I tested that.
I'll fix this.

Comment (dellagustin) [reply server/index.js:130]: Having some trouble running my local setup now: https://github.com/Podcastindex-org/web-ui/issues/231

Comment (dellagustin) [reply server/index.js:130]: So, it turns out I made two mistakes here that cancelled each other, so it still works out.
It still works when `socialInteract `is there and with one or more items with `activitypub`.
So, it is not urgent.
I'll transfer to an issue and take care later.

Comment (dellagustin) [reply server/index.js:130]: There you go, https://github.com/Podcastindex-org/web-ui/issues/232.
I'll follow up from there and resolve this comment.
